pkg/kubelet/kuberuntime/kuberuntime_manager.go

```diff
@@ -474,6 +474,17 @@ func (m *kubeGenericRuntimeManager) computePodActions(pod *v1.Pod, podStatus *ku
 	// check the status of containers.
 	for idx, container := range pod.Spec.Containers {
 		containerStatus := podStatus.FindContainerStatusByName(container.Name)
+
+		// Call internal container post-stop lifecycle hook for any non-running container so that any
+		// allocated cpus are released immediately. If the container is restarted, cpus will be re-allocated
+		// to it.
+		if containerStatus != nil && containerStatus.State != kubecontainer.ContainerStateRunning {
```

Comment:
what will happen when `containerStatus.State == kubecontainer.ContainerStateCreated` ?

Reply:
The container will be restarted as per https://github.com/kubernetes/kubernetes/blob/ccaaf5cad5901f26283bfa8c2b230dc1dc3579b4/pkg/kubelet/container/helpers.go#L77. See https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kuberuntime/kuberuntime_manager.go#L480.

Reply:
The same question could be asked for the existing condition in the next block. I believe this is because in the runtime, `startContainer` both creates and starts the container so we won't see the created state from this sync loop.

Reply:
To be more explicit, we could check for only the terminal container states. WDYT @balajismaniam ?

Reply:
It would be even nicer if there were a helper predicate to determine which container states are actually terminal. Then if a new one is added we wouldn't have to track down and re-vet all the conditions. Nice to have, we can propose that in a follow-up maybe.

Reply:
@ConnorDoyle I think we should leave the code as is (i.e., check for all non-running container states). It will help in a scenarios where `createContainer` + pre-start lifecycle hook is called but not `startContainer` (might not be possible but it will be useful to guard against it). Also, if the container ends-up in create or unknown state, it will be restarted anyways.

Reply:
Agree with the helper predicate. We can propose it.

Reply:
is it possible that this hook gets called multiple times for the same container?

Reply:
Yes, it is possible. But the call is idempotent and hence safe to call here.

Reply:
Yes, this is already the case, since we call RemoveContainer from both PreStopContainer and PostStopContainer.